posts/2023-12-21-async-fn-rpit-in-traits.md

```diff
@@ -0,0 +1,229 @@
+---
+layout: post
+title: "Announcing `async fn` and return-position `impl Trait` in traits"
+author: Tyler Mandry
+team: The Async Working Group <https://www.rust-lang.org/governance/wgs/wg-async>
+---
+
+The Rust Async Working Group is excited to announce major progress towards our goal of enabling the use of `async fn` in traits. The next Rust release will include stable support for both `-> impl Trait` notation and `async fn` in traits.
+
+This is a big milestone, and we know many users will be itching to try these out in their own code. However, we are still missing some important features that many users need. Read on for an recommendations on when and how to use the stabilized features.
+
+## What's stabilizing
+
+Ever since [RFC #1522], Rust has allowed users to write `impl Trait` as the return type of functions. This means that the function returns "some type that implements `Trait`". This is commonly used to return iterators, closures, or other types that are complex or even impossible to write explicitly:
+
+[RFC #1522]: https://rust-lang.github.io/rfcs/1522-conservative-impl-trait.html
+
+```rust
+/// Given a list of players, return an iterator
+/// over their names.
+fn player_names(
+    players: &[Player]
+) -> impl Iterator<Item = &String> {
+    players
+        .iter()
+        .map(|p| &p.name)
+}
+```
+
+Starting in Rust 1.75, which hits stable next week, you can use **return-position `impl Trait` in traits** (RPITIT) and trait impls. For example, you could use this to write a trait that returns an iterator:
+
+```rust
+trait Container {
+    fn items(&self) -> impl Iterator<Item = Widget>;
+}
+
+impl Container for MyContainer {
+    fn items(&self) -> impl Iterator<Item = Widget> {
+        self.items.iter().cloned()
+    }
+}
+```
+
+So what does all of this have to do with async functions? Well, async functions are "just sugar" for a function that returns `-> impl Future`. Since that is now permitted in traits, **we also permit you to write traits that use `async fn`**.
+
+```rust
+trait HttpService {
+    async fn fetch(&self, url: Url) -> HtmlBody;
+//  ^^^^^^^^ desugars to:
+//  fn fetch(&self, url: Url) -> impl Future<Output = HtmlBody>;
+}
+```
+
+## Where the gaps lie
+
+### `-> impl Trait` in public traits
+
+The use of `-> impl Trait` is still discouraged for general use in **public** traits, for the reason that users can't use additional bounds on the return type. For example, there is no way to write this function:
```

Comment:
```suggestion
The use of `-> impl Trait` is still discouraged for general use in **public** traits for the reason that implementors of the trait cannot provide additional bounds on the opaque return types that can be relied on by users. For example, there is no way to write this function:
```

We're making a really subtle point here, and we do want to keep it concise, but it perhaps may be good to spell this out a bit further.

Reply:
I am not sure I agree that this clarifies it. Implementers can provide additional bounds, either through refinement or auto trait leakage, and users can rely on them _if_ they know the type being used that implements the trait. The issue is that generic code which does not know the type cannot state additional requirements which it can then rely on.

Reply:
Made some clarifications.

Reply:
Fair enough.  Earlier drafts of my suggestion did get into this but were too long.  I agree it's OK as it is.

It's a subtle point and I hope that people get what we mean.

Reply:
We were apparently typing at the same time, and GH didn't show me the clarifications prior to my message above.  The clarifications look good.